File: MultitenancyDatabase/src/main/java/br/com/devcave/multitenancy/controller/ProductController.java
package br.com.devcave.multitenancy.controller;

import br.com.devcave.multitenancy.domain.Product;
import br.com.devcave.multitenancy.repository.ProductRepository;
import io.swagger.annotations.ApiImplicitParam;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpEntity;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.servlet.support.ServletUriComponentsBuilder;

import java.net.URI;
import java.util.List;

@RestController
@RequiredArgsConstructor
@RequestMapping("product")
public class ProductController {

    private final ProductRepository productRepository;

    @PostMapping
    @ApiImplicitParam(name = "tenantId", dataType = "string", paramType = "header")
    public HttpEntity<?> create(@RequestBody final Product product) {
        Product productSaved = productRepository.save(product);
        URI location = ServletUriComponentsBuilder
                .fromCurrentRequest()
                .path("/{id}")
                .build(productSaved.getId());

        return ResponseEntity.created(location).build();
    }

    @GetMapping("{id}")
    @ApiImplicitParam(name = "tenantId", dataType = "string", paramType = "header")
    public HttpEntity<Product> findById(@PathVariable final Long id) {
        Product product = productRepository.findById(id).orElse(null);

        return ResponseEntity.ok(product);
    }

    @GetMapping
    @ApiImplicitParam(name = "tenantId", dataType = "string", paramType = "header")
    public HttpEntity<List<Product>> findAll() {
        List<Product> productList = productRepository.findAll();

        return ResponseEntity.ok(productList);
    }
}

File: MultitenancyDiscriminator/src/main/kotlin/br/com/devcave/multitenancy/controller/ProductController.kt
package br.com.devcave.multitenancy.controller

import br.com.devcave.multitenancy.domain.Product
import br.com.devcave.multitenancy.domain.ProductRequest
import br.com.devcave.multitenancy.service.ProductService
import io.swagger.v3.oas.annotations.Parameter
import io.swagger.v3.oas.annotations.enums.ParameterIn
import io.swagger.v3.oas.annotations.responses.ApiResponse
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("products")
class ProductController(
    private val productService: ProductService
) {
    @GetMapping("{id}")
    @Parameter(name = "tenantId", `in` = ParameterIn.HEADER)
    fun findById(@PathVariable id:Long): Product {
        return productService.findById(id)
    }

    @GetMapping
    @Parameter(name = "tenantId", `in` = ParameterIn.HEADER)
    fun findAll(): List<Product> {
        return productService.findAll()
    }

    @GetMapping("name")
    @Parameter(name = "tenantId", `in` = ParameterIn.HEADER)
    fun findByName(@RequestParam name:String): List<Product> {
        return productService.findByName(name)
    }

    @PostMapping
    @Parameter(name = "tenantId", `in` = ParameterIn.HEADER)
    fun create(request: ProductRequest): Long {
        return productService.save(request)
    }

    @DeleteMapping("{id}")
    @Parameter(name = "tenantId", `in` = ParameterIn.HEADER)
    fun deleteById(@PathVariable id: Long) {
        productService.deleteById(id)
    }
}